exetera/processing/date_time_helpers.py

```diff
@@ -0,0 +1,172 @@
+from typing import Sequence, Tuple, Union
+from datetime import datetime, timedelta
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+from exetera.core.utils import SECONDS_PER_DAY
+
+
+
+def get_periods(start_date: datetime,
+                end_date: datetime,
+                period: str,
+                delta: int = 1
+                ) -> Sequence[datetime]:
+    """
+    Generate a set of periods into which timestamped data can be grouped.
+    Delta controls whether the sequence of periods is generated from an start point
+    or an end point. When delta is positive, the sequence is generated forwards in time.
+    When delta is negative, the sequence is generate backwards in time.
+    :param start_date: a ``datetime.datetime`` object for the starting period
+    :param end_date: a ``datetime.datetime`` object for tne ending period, exclusive
+    :param period: a string representing the unit in which the delta is calculated
+    ('day', 'days', 'week', 'weeks')
+    :param delta: an integer representing the delta.
+    :return:
+    """
+
+    period_map = {
+        'day': lambda x: timedelta(days=x),
+        'days': lambda x: timedelta(days=x),
+        'week': lambda x: timedelta(weeks=x),
+        'weeks': lambda x: timedelta(weeks=x)
+    }
+    if not isinstance(period, str):
+        raise ValueError("'period' must be of type str but is {}".format(type(period)))
+    if period not in period_map.keys():
+        raise ValueError("'period': must be one of {} but is {}".format(period_map.keys(), period))
+    if not isinstance(delta, int):
+        raise ValueError("'delta': element 1 must of type int but is {}".format(type(delta)))
+    if delta == 0:
+        raise ValueError("'delta' cannot be 0")
+    if delta < 0:
+        if start_date < end_date:
+            raise ValueError("'start_date' must be greater than 'end_date' if 'delta' is negative")
+    else:
+        if start_date > end_date:
+            raise ValueError("'start_date' must be less than 'end_date' if 'delta' is positive")
+
+    tdelta = period_map[period](delta)
+    if delta > 0:
+        dates = [start_date]
+        cur_date = start_date + tdelta
+        while cur_date <= end_date:
+            dates.append(cur_date)
+            cur_date += tdelta
+    else:
+        dates = [start_date]
+        cur_date = start_date + tdelta
+        while cur_date >= end_date:
+            dates.append(cur_date)
+            cur_date += tdelta
+    return dates
+
+
+def get_days(date_field: ArrayLike[np.float64],
+             date_filter: ArrayLike[bool] = None,
+             start_date: np.float64 = None,
```

Comment:
This needs to be type as `Optional[np.float64]`, I think this is what the test fail is referring to.

Reply:
It wasn't what the test fail was referring to; that was attempting to use ArrayLike with a template. I have changed the start and end dates to be Optional[np.float64] however